packages/backend/src/handlers/actions/3Box/refresh3BoxCache/cache3BoxProfiles.ts

```diff
@@ -0,0 +1,16 @@
+import { CacheProcessOutput } from '../../../../lib/autogen/hasura-sdk';
+import { client } from '../../../../lib/hasuraClient';
+import { updateCachedProfile } from '../cache3BoxProfile/updateCachedProfile';
+
+export async function cache3BoxProfiles(): Promise<CacheProcessOutput> {
+  const data = await client.GetPlayerIds();
+  const ids = data.player.map(({ id }) => id);
+
+  try {
+    await Promise.all(ids.map((id) => updateCachedProfile(id)));
```

Comment:
this would trigger all these requests to happen in parallel rather than one at a time, which would spike CPU/memory usage and also result in timeouts from the API getting hammered with hundreds of concurrent requests.  once we start growing out user base, this wont scale

If we want to iterate over all of them, we should use something like https://github.com/sergelerator/async-batch to limit the concurrency to 5-10 players at a time.

Also, this method of periodically fetching all user data would result in things getting out of sync (if period is too long), or over-fetching unnecessarily and using up a lot of resources (if period is frequent). The ideal strategy to minimize the chance of stale data + prevent having to fetch every single user periodically would be to return the cached data when its queried + trigger an update in the background so that next time someone visits that same player page they get fresh data. For players that no one visits, we don't waste time updating the data frequently and can do it once every 3-4 days instead.

It would also be nice to store a timestamp in the table of when the cached data was last updated and only fetch data from IDX when its X days stale. Basically a TTL.

Reply:
I used [Bottleneck](https://github.com/SGrondin/bottleneck#readme) with a `maxConcurrency` of 10 to rate limit the update processes. It is stored as a shared variable on the Express server, and there's a map for tracking its current contents.

There's a `last_checked_at` field on the cache entries now and a Hasura action that invalidates all entries more than four days old meant to be run from a cron job.